cluster/v1alpha1/types.go

```diff
@@ -0,0 +1,100 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope="Cluster"
+
+// ManagedClusterSet is a set of managed clusters.
+type ManagedClusterSet struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the attributes of the desired managed clusters
+	Spec ManagedClusterSetSpec `json:"spec"`
+
+	// Status represents the current status of the cluster set
+	// +optional
+	Status ManagedClusterSetStatus `json:"status,omitempty"`
+}
+
+// ManagedClusterSetSpec describes the attributes of the desired managed clusters
+type ManagedClusterSetSpec struct {
+	// ClusterSelectors represents a slice of selectors to select managed clusters
+	// If empty, the cluster set will include all managed clusters.
+	// If more than one ClusterSelector are specified in the slice, OR operation
+	// will be used between them.
+	// +optional
+	ClusterSelectors []ClusterSelector `json:"clusterSelectors,omitempty"`
+}
+
+// ClusterSelector represents a selector of managed clusters
+// ClusterNames and LabelSelector are mutually exclusive. They cannot be set at the
+// same time. If none of them is set, all managed clusters will be selected
+type ClusterSelector struct {
+	// ClusterNames represents a list of cluster name
+	// +optional
+	ClusterNames []string `json:"clusterNames,omitempty"`
+
+	// LabelSelector represents a label selector to select cluster by label
+	// +optional
+	LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty"`
+}
+
+// ManagedClusterSetStatus represents the current status of the cluster set.
+type ManagedClusterSetStatus struct {
+	// ClusterNames contains names of all the selected clusters.
+	// +optional
+	ClusterNames []string `json:"clusterNames,omitempty"`
```

Comment:
I wonder if we really want to reflect this in status. On the surface it absolutely makes sense, I wonder about how it would scale in storage:

- 1MB upper limit on etcd records means a soft upper limit on number of clusters that can be represented in the entire object
- What happens if you select too many clusters to write into the status?

Reply:
Removed from status.